### PR TITLE
Python3 update for meta-freescale layer

### PR DESCRIPTION
**Pull request by zandrey**
Python2 support has been recently dropped from OE-Core due to it's effective EOL, therefore some recipes and corresponding source code packages should be updated to use Python3 BitBake classes.

@otavio @thochstein:
`optee-imx` and `imx-gpu-apitrace` has source code patches included in order to be able to use Python3 and should be applied at the source. Can you please forward them so they could be removed from the layer afterwards?

Thanks a lot!

-- andrey

### CODE REVIEW

File: recipes-graphics/imx-gpu-apitrace/imx-gpu-apitrace/0002-cmake-Use-PythonInterp-instead-of-Python3.patch [hunk 1]
```diff
@@ -0,0 +1,225 @@
+From 762587079dea1c24de95c38c271ed1692c8f2a18 Mon Sep 17 00:00:00 2001
+From: Jose Fonseca <jfonseca@vmware.com>
+Date: Sun, 14 Apr 2019 22:22:24 +0900
+Subject: [PATCH 2/7] cmake: Use PythonInterp instead of Python3.
+
+As the latter is not in widely available CMake versions.
+
+Upstream-Status: Pending
```
Comment (otavio) [inline recipes-graphics/imx-gpu-apitrace/imx-gpu-apitrace/0002-cmake-Use-PythonInterp-instead-of-Python3.patch]: It seems to be a backport, isn't it?

File: recipes-graphics/imx-gpu-apitrace/imx-gpu-apitrace/0004-cli-Pickle-blobs-as-Python-3-bytes-objects.patch [hunk 2]
```diff
@@ -0,0 +1,82 @@
+From 63b651332cfeb02cc723675ad2d150641d037e21 Mon Sep 17 00:00:00 2001
+From: Jose Fonseca <jfonseca@vmware.com>
+Date: Sun, 28 Apr 2019 13:47:29 +0100
+Subject: [PATCH 4/7] cli: Pickle blobs as Python 3 bytes objects.
+
+Upstream-Status: Pending
```
Comment (otavio) [inline recipes-graphics/imx-gpu-apitrace/imx-gpu-apitrace/0004-cli-Pickle-blobs-as-Python-3-bytes-objects.patch]: Here too

File: recipes-graphics/imx-gpu-apitrace/imx-gpu-apitrace/0002-cmake-Use-PythonInterp-instead-of-Python3.patch (hunk 1 above)
Comment (zandrey) [reply recipes-graphics/imx-gpu-apitrace/imx-gpu-apitrace/0002-cmake-Use-PythonInterp-instead-of-Python3.patch]: Correct! This one is ported, would drop it now.

File: recipes-graphics/imx-gpu-apitrace/imx-gpu-apitrace/0004-cli-Pickle-blobs-as-Python-3-bytes-objects.patch (hunk 2 above)
Comment (zandrey) [reply recipes-graphics/imx-gpu-apitrace/imx-gpu-apitrace/0004-cli-Pickle-blobs-as-Python-3-bytes-objects.patch]: I've checked the latest state of `imx_7.1` branch and looks like this patch applies. It modifies generated blobs from using _bytearrays_ to python3 _byte objects_.

Original support for _bytearrays_ has been introduced with eba2deca7 from upstream, and is not changed at the HEAD of `imx_7.1`.

**Comment by thochstein** [comment]
> `optee-imx` and `imx-gpu-apitrace` has source code patches included in order to be able to use Python3 and should be applied at the source. Can you please forward them so they could be removed from the layer afterwards?

Will do, thanks a lot Andrey!

**Comment by zandrey** [comment]
@thochstein Thanks for taking those in!

Can you please wait with forwarding those patches, @otavio has correctly pointed out that some of patches in `imx-gpu-apitrace` were already backported and are not required.

I'm working on this now to sort them out, and would update the PR with only those that are necessary and not backported yet.

Thanks a lot!

-- andrey

File: recipes-graphics/imx-gpu-apitrace/imx-gpu-apitrace/0004-cli-Pickle-blobs-as-Python-3-bytes-objects.patch (hunk 2 above)
Comment (zandrey) [reply recipes-graphics/imx-gpu-apitrace/imx-gpu-apitrace/0004-cli-Pickle-blobs-as-Python-3-bytes-objects.patch]: I thought about this a bit, and actually this patch is not really crucial here to have python3 support. Originally, it has been introduced on the `python3` branch in upstream together with all other modifications required to move to python3, that's what made me think this one is also necessary.

I've tried to exclude this patch from recipe and build succeeded.

@otavio @thochstein Can you please advise whether you'd prefer this patch should be included in the series or should I drop it?

Thanks a lot!

-- andrey

Comment (zandrey) [reply recipes-graphics/imx-gpu-apitrace/imx-gpu-apitrace/0004-cli-Pickle-blobs-as-Python-3-bytes-objects.patch]: I've dropped this patch as well and updated PR with new version.

**Comment by zandrey** [comment]
@otavio Got the `Upstream-Status` set correctly with links to commits in the upstream. I guess now it should be OK.

-- andrey

**Comment by zandrey** [comment]
@otavio Thanks for taking this in! Now the layer can be built with latest oe-core (at least the python3-related recipes).

@thochstein Can you please examine patches to `optee-os` and `imx-gpu-apitrace` and send them internally? I've committed only those that are required to resolve python3 build issues, they are ready to be forwarded.

Thank you all!

-- andrey

**Comment by thochstein** [comment]
> @thochstein Can you please examine patches to `optee-os` and `imx-gpu-apitrace` and send them internally? I've committed only those that are required to resolve python3 build issues, they are ready to be forwarded.

Thanks, @zandrey. We'll work on this for our 5.4 release end of March, so we'll contribute to meta-freescale in April. We'll bump apitrace to 9.0 and add your optee-imx patch.

**Comment by thochstein** [comment]
> Thanks, @zandrey. We'll work on this for our 5.4 release end of March, so we'll contribute to meta-freescale in April. We'll bump apitrace to 9.0 and add your optee-imx patch.

We are working on optee version 3.7, which already has python 3 support. We'd rather not modify our current 3.2 source.

**Comment by zandrey** [comment]
> 
> We are working on optee version 3.7, which already has python 3 support. I checked your patch on our 3.7 and found it to be an exact match (except for 2 minor whitespace changes).
> 
> We'd rather not modify our current 3.2 source.

@thochstein That's OK, as long as the patch is in the layer - `master` could be built for 4.19.y kernel with optee 3.2 and @otavio is fine with keeping it here.

It was actually my task to report the patch upstream (OE rules of the game), but if upstream decides it is better not to take it - we would have it applied at the meta layer level.

Thanks a lot for the feedback!

-- andrey

**Pull request merged**